src/BinCommand.php

```diff
@@ -33,57 +34,85 @@ protected function configure()
      */
     public function execute(InputInterface $input, OutputInterface $output)
     {
-        $this->resetComposers();
+        $this->resetComposers($application = $this->getApplication());
+        /** @var ComposerApplication $application */
 
         putenv('COMPOSER_BIN_DIR='.$this->createConfig()->get('bin-dir'));
 
-        $binVendorRoot = 'vendor-bin';
+        $binvendorRoot = 'vendor-bin';
         $binNamespace = $input->getArgument('namespace');
         $input = new StringInput(preg_replace('/bin\s+' . preg_quote($binNamespace, '/') . '/', '', $input->__toString(), 1));
 
-        if ('all' !== $binNamespace) {
-            $binRoot = $binVendorRoot.'/'.$binNamespace;
-            if (!file_exists($binRoot)) {
-                mkdir($binRoot, 0777, true);
-            }
-
-            $this->chdir($binRoot);
-
-            return $this->getApplication()->doRun($input, $output);
-        }
+        return ('all' !== $binNamespace)
+            ? $this->executeInNamespace($application, $binvendorRoot.'/'.$binNamespace, $input, $output)
+            : $this->executeAllNamespaces($application, 'vendor-bin', $input, $output)
+        ;
+    }
 
+    /**
+     * @param ComposerApplication $application
+     * @param                     $binVendorRoot
+     * @param InputInterface      $input
+     * @param OutputInterface     $output
+     *
+     * @return int
+     */
+    private function executeAllNamespaces(ComposerApplication $application, $binVendorRoot, InputInterface $input, OutputInterface $output)
+    {
         $binRoots = glob($binVendorRoot.'/*', GLOB_ONLYDIR);
         if (empty($binRoots)) {
             $this->getIO()->writeError('<warning>Couldn\'t find any bin namespace.</warning>');
 
-            return;
+            return 1;
```

Comment:
before nothing was returned, but in this case isn't it better to return `1` to say that an error occurred?